src/libcore/iter/traits.rs

```diff
@@ -581,13 +582,15 @@ pub trait Product<A = Self>: Sized {
     fn product<I: Iterator<Item=A>>(iter: I) -> Self;
 }
 
+// NB: explicitly use Add and Mul here to inherit overflow checks
 macro_rules! integer_sum_product {
     ($($a:ident)*) => ($(
         #[stable(feature = "iter_arith_traits", since = "1.12.0")]
         impl Sum for $a {
+            #[rustc_inherit_overflow_checks]
             fn sum<I: Iterator<Item=$a>>(iter: I) -> $a {
                 iter.fold(0, |a, b| {
-                    a.checked_add(b).expect("overflow in sum")
+                    Add::add(a, b)
```

Comment:
This is not what I meant, although it would work (even without any added `rustc_inherit_overflow_checks`).
The intention was to replace the closure, i.e. `iter.fold(0, Add::add)`.